simdev/module/github/github_api_wrapper.py

```diff
@@ -0,0 +1,209 @@
+        :param api_tokens: list of GitHub API tokens
+        :param github_api_url: URL of GitHub API
+        :param max_retries_num: max number of retries
+        fetching content from GitHub
+        """
+        self._api_url = github_api_url
+        self._max_retries_num = max_retries_num
+        self._headers: Dict[str, str] = {
+            "Accept": "application/vnd.github+json",
+            "X-GitHub-Api-Version": "2022-11-28",
+        }
+        self._api_tokens = api_tokens
+        self._current_api_token: str | None = None
+        self._update_api_token()
+
+        # Making these functions get data from cache if possible
+        self.fetch_stargazers = PipelineCache.memory.cache(
+            func=self.fetch_stargazers, ignore=["progress", "batch_count", "page_limit"]
+        )
+        self.fetch_starred_repositories = PipelineCache.memory.cache(
+            func=self.fetch_starred_repositories,
+            ignore=["progress", "batch_count", "page_limit"],
+        )
+
+    def fetch_stargazers(
+            self, repo: str, batch_count=100, page_limit=400,
+            progress: tqdm.tqdm = None
+    ) -> Set[str]:
+        """
+        Fetch stargazers of the repository
+        :param repo: name (GitHub flavoured. Example: theseems/HseNotebooks)
+        :param batch_count: entries per GitHub API response page
+        :param page_limit: max amount of pages to fetch
+        :param progress: TQDM progress if any
+        :return: set of stargazers
+        """
+        if progress is not None:
+            progress.desc = f"Fetching stargazers for {repo}"
+
+        fixed_url_part = (
+            f"{self._api_url}repos/{repo}/stargazers?per_page={str(batch_count)}&page="
+        )
+        current_page = 0
+        result: Set[str] = set()
+
+        while True:
+            full_url = F"{fixed_url_part}{current_page}"
+            _update_tqdm_page_postfix(progress, current_page + 1)
+            page_response: List[Dict[str, Any]] = self._get_request_json(full_url)
+            if len(page_response) == 0:
+                break
+            if not isinstance(page_response, list):
+                _report_unknown_response(page_response, full_url)
+                continue
+            for stargazer in page_response:
+                result.add(stargazer["login"])
+            current_page += 1
+            if current_page > page_limit:
+                break
+        return result
+
+    def fetch_starred_repositories(
+            self, user: str, batch_count=100, page_limit=400,
+            progress: tqdm.tqdm = None
+    ) -> Set[str]:
+        """
+        Fetch repositories starred by a user
+        :param user: to fetch starred repositories of
+        :param batch_count: number of entries per GitHub API response's page
+        :param page_limit: max amount of pages to fetch
+        :param progress: TQDM progress
+        :return: set of starred repositories
+        """
+        if progress is not None:
```

Comment:
make it `update_progress_bar_descriptor` - you are using it here and in the function above